pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -245,6 +327,16 @@ func validatePipelineWorkspaces(wss []WorkspacePipelineDeclaration, pts []Pipeli
 			}
 		}
 	}
+	for tIdx, t := range finalTasks {
+		for wsIdx, ws := range t.Workspaces {
+			if _, ok := wsTable[ws.Workspace]; !ok {
+				return apis.ErrInvalidValue(
+					fmt.Sprintf("pipeline task %q expects workspace with name %q but none exists in pipeline spec", t.Name, ws.Workspace),
+					fmt.Sprintf("spec.finally[%d].workspaces[%d]", tIdx, wsIdx),
+				)
+			}
+		}
+	}
```

Comment:
hmmm, is it possible to share any of this logic b/w v1alpha1 and v1beta1? using separate logic for results led to some serious bugs and confusion so i think it's worth investing some time into finding ways for them both to call the exact same code

(also a great chance to break up some of this validation with unit tests, vs testing it all by passing in complete pipelines)

this is a big ask but also a good candidate for a separate PR that could be merged before this one?

Reply:
I am sure it is possible to share logic between v1alpha1 and v1beta1 also love your idea of breaking this validation and testing them piece by piece instead of creating complete pipeline.

I am trying to think about possible fix here, one option could be keep the local functions to validate graph, workspaces, resources, etc in `v1beta1` and make them public so that the same functions can be used by `v1alpha1` after converting `v1alpha1` resources to `v1beta1` (mainly `PipelineTask`) 🤭  

Also, looked at moving those local functions to a common utility 🤔

Reply:
something like [this](https://github.com/tektoncd/pipeline/compare/master...pritidesai:refactor-pipeline-validation) 🤔

Reply:
Have done huge refactoring by separating positive tests and negative tests, also by renaming test package to `v1beta1` to test individual functions, will create validation routine as part of `PipelineTask` like @bobcatfish mentioned here: https://github.com/tektoncd/pipeline/pull/2650/files#r428820876